Specify Bloom Indexing (#6)
fission-suite/spec: car-pool/car-mirror/SPEC.md

└─────────────────────┘   │                           │        └───────────────┘
```

## 3.4 Bloom Filter

### 3.4.1 Indexing

> Indexes generated from hashes MUST follow the following strategy based on if the filter fits perfectly into a power of two ($2^c$). This is a single algorithm, but if the size of the Bloom filter is a power of two, rejection sampling MAY be omitted. Using a Bloom filter that is a power of two is RECOMMENDED since it avoids resampling.

Reviewer: ```suggestion
Indexes MUST be generated from hashes using the following strategy, based on whether or not the filter fits perfectly into a power of two ($2^c$). This is a single algorithm, but if the size of the Bloom filter is a power of two, rejection sampling MAY be omitted. Using a Bloom filter that is a power of two is RECOMMENDED since it avoids resampling.
```

> All hashes generated for indexing MUST be interpreted as big endian.

Reviewer: ```suggestion
All hashes generated for indexing MUST be interpreted as big-endian.
```

Reviewer: I think we need a comment on the binary representation of the filter itself also being big-endian as well.

Author: You mean the indexes? Will add!

Reviewer: I mean the actual bloom bytes that get sent over the wire (i.e. `bb` in the payload).  Or maybe that is just a given, due to how CBOR encodes bytes?  See "big-endian" in https://datatracker.ietf.org/doc/html/rfc7049.

Reviewer: My point is just that when bytes are sent over the wire we want to ensure they are being handled the same way on both ends.

#### 3.4.1.1 Power of Two

If the size $m$ of the filter is $d$ powers of two ($2^d$), take the lowest (rightmost) $d$ bits from the hash and interpret it as an index.

#### 3.4.1.2 Rejection Sampling

A Bloom filter MAY be a length that is not a power of two. This is NOT RECOMMENDED since it incurs [rejection sampling](https://en.wikipedia.org/wiki/Rejection_sampling) overhead.

> This case uses the nearest rounded power of two as in [3.4.1.1](#3411-power-of-two). If the sampled number is less than $m$, then use it MUST be used as the index. If the number is larger, then the right shift the unmasked number, take the required number of bits (e.g. via AND-mask) and check again. Repeat this process until the number of digits is exhausted.

Reviewer: ```suggestion
This case uses the nearest rounded power of two as in [3.4.1.1](#3411-power-of-two). If the sampled number is less than $m$, then it MUST be used as the index. If the number is larger, then right shift the unmasked number, take the required number of bits (e.g. via AND-mask) and check again. Repeat this process until the number of digits is exhausted.
```

> If none of the samples succeeds, a new hash MUST be generated and this process begun again. Hash generation MUST be performed via [XXH3](https://cyan4973.github.io/xxHash/) with the rehashing generation used as a seed (a simple counter).

Reviewer: ```suggestion
If none of the samples succeed, a new hash MUST be generated and this process begun again. Hash generation MUST be performed via [XXH3](https://cyan4973.github.io/xxHash/) with the rehashing generation used as a seed (a simple counter).
```
